File: src/YandexDiskSDK/YandexDiskSDK/Interfaces/SharingClient.vb
Imports Newtonsoft.Json
Imports YandexDiskSDK.JSON
Imports YandexDiskSDK.utilitiez

Public Class SharingClient
    Implements ISharing



#Region "ListAllSharedLinks"
    Public Async Function Get_ListAllSharedLinksOfUser(Optional Target As ItemTypeEnum = ItemTypeEnum.both, Optional Fields As Fields = Fields.nothing, Optional PreviewSize As PreviewSizeEnum = PreviewSizeEnum.S_150, Optional Limit As Integer = 20, Optional Offset As Integer = 0) As Task(Of JSON_FolderList) Implements ISharing.ListAllSharedLinks
        Dim parameters = New Dictionary(Of String, String)
        If Not Fields = Fields.items OrElse Fields = Fields.type Then parameters.Add("fields", Fields.ToString)
        parameters.Add("limit", Limit)
        parameters.Add("offset", Offset)
        parameters.Add("preview_size", stringValueOf(PreviewSize))
        If Not Target = ItemTypeEnum.both Then parameters.Add("type", Target.ToString)

        Using localHttpClient As New HttpClient(New HCHandler)
            Dim RequestUri = New pUri("resources/public", parameters)
            Using response As Net.Http.HttpResponseMessage = Await localHttpClient.GetAsync(RequestUri).ConfigureAwait(False)
                Dim result As String = Await response.Content.ReadAsStringAsync()

                If response.StatusCode = Net.HttpStatusCode.OK Then
                    Dim fin As New JSON_FolderList
                    fin.limit = result.Jobj.SelectToken("limit").ToString
                    fin.offset = result.Jobj.SelectToken("offset").ToString
                    fin._Files = (From c In result.Jobj.SelectToken("items").ToList().Select(Function(i, JSON_FileMetadata) i).Where(Function(i) i.SelectToken("type").ToString = "file").Select(Function(i) JsonConvert.DeserializeObject(Of JSON_FileMetadata)(i.ToString, JSONhandler))).ToList()
                    fin._Folders = (From c In result.Jobj.SelectToken("items").ToList().Select(Function(i, JSON_FolderMetadata) i).Where(Function(i) i.SelectToken("type").ToString = "dir").Select(Function(i) JsonConvert.DeserializeObject(Of JSON_FolderMetadata)(i.ToString, JSONhandler))).ToList()
                    Return fin
                Else
                    ShowError(result)
                End If
            End Using
        End Using
    End Function
#End Region

#Region "PublicFolderContents"
    Public Async Function _ParsePublicFolder2(FolderPublicUrl As Uri, Optional FilePathInsidePublicFolder As String = Nothing, Optional Sort As SortEnum? = Nothing, Optional PreviewCrop As Boolean = True, Optional PreviewSize As PreviewSizeEnum = PreviewSizeEnum.S_150, Optional Limit As Integer = 20, Optional Offset As Integer = 0) As Task(Of JSON_PublicFolder) Implements ISharing.PublicFolderContents
        FolderPublicUrl.Validation(ItemTypeEnum.dir)
        Return Await _ParsePublicFolder(FolderPublicUrl.ToString, FilePathInsidePublicFolder, Sort, PreviewCrop, PreviewSize, Limit, Offset)
    End Function
    Public Async Function _ParsePublicFolder(FolderPublicKey As String, Optional FilePathInsidePublicFolder As String = Nothing, Optional Sort As SortEnum? = Nothing, Optional PreviewCrop As Boolean = True, Optional PreviewSize As PreviewSizeEnum = PreviewSizeEnum.S_150, Optional Limit As Integer = 20, Optional Offset As Integer = 0) As Task(Of JSON_PublicFolder) Implements ISharing.PublicFolderContents
        Dim parameters = New Dictionary(Of String, String)
        parameters.Add("public_key", System.Net.WebUtility.UrlEncode(FolderPublicKey))
        parameters.Add("path", FilePathInsidePublicFolder)
        parameters.Add("limit", Limit)
        parameters.Add("offset", Offset)
        parameters.Add("preview_crop", PreviewCrop)
        parameters.Add("preview_size", stringValueOf(PreviewSize))
        If Sort.HasValue Then parameters.Add("sort", Sort.ToString)

        Using localHttpClient As New HttpClient(New HCHandler)
            Dim RequestUri = New pUri("public/resources", parameters)
            Using response As Net.Http.HttpResponseMessage = Await localHttpClient.GetAsync(RequestUri).ConfigureAwait(False)
                Dim result As String = Await response.Content.ReadAsStringAsync()

                If response.StatusCode = Net.HttpStatusCode.OK Then
                    Dim fin = JsonConvert.DeserializeObject(Of JSON_PublicFolder)(result, JSONhandler)
                    fin.ItemsList._Files = (From c In result.Jobj.SelectToken("_embedded")("items").ToList().Select(Function(i, JSON_FileMetadata) i).Where(Function(i) i.SelectToken("type").ToString = "file").Select(Function(i) JsonConvert.DeserializeObject(Of JSON_FileMetadata)(i.ToString, JSONhandler))).ToList()
                    fin.ItemsList._Folders = (From c In result.Jobj.SelectToken("_embedded")("items").ToList().Select(Function(i, JSON_FolderMetadata) i).Where(Function(i) i.SelectToken("type").ToString = "dir").Select(Function(i) JsonConvert.DeserializeObject(Of JSON_FolderMetadata)(i.ToString, JSONhandler))).ToList()
                    Return fin
                Else
                    ShowError(result)
                End If
            End Using
        End Using
    End Function
#End Region

#Region "GetDownloadUrlOfFileInPublicFolder"
    Public Async Function _GetDownloadUrl(FolderPublicUrl As Uri, FilePathInsidePublicFolder As String) As Task(Of String) Implements ISharing.GetDownloadUrlOfFileInPublicFolder
        FolderPublicUrl.Validation(ItemTypeEnum.dir)
        Return Await POST_GetDownloadUrl(FolderPublicUrl.ToString, FilePathInsidePublicFolder)
    End Function
    Public Async Function POST_GetDownloadUrl(FolderPublicKey As String, FilePathInsidePublicFolder As String) As Task(Of String) Implements ISharing.GetDownloadUrlOfFileInPublicFolder
        Dim parameters = New Dictionary(Of String, String)
        parameters.Add("public_key", System.Net.WebUtility.UrlEncode(FolderPublicKey))
        parameters.Add("path", FilePathInsidePublicFolder)

        Using localHttpClient As New HttpClient(New HCHandler)
            Dim RequestUri = New pUri("resources/download", parameters)
            Using response As Net.Http.HttpResponseMessage = Await localHttpClient.GetAsync(RequestUri).ConfigureAwait(False)
                Dim result As String = Await response.Content.ReadAsStringAsync()

                If response.StatusCode = Net.HttpStatusCode.OK Then
                    Return result.Jobj.SelectToken("href").ToString
                Else
                    ShowError(result)
                End If
            End Using
        End Using
    End Function
#End Region

#Region "PublicUrlToPublicKey"
    Public Async Function _PublicUrlToPublicKey(PublicUrl As Uri) As Task(Of String) Implements ISharing.PublicUrlToPublicKey
        PublicUrl.Validation(ItemTypeEnum.both)
        Dim parameters = New Dictionary(Of String, String)
        parameters.Add("public_key", System.Net.WebUtility.UrlEncode(PublicUrl.ToString))
        parameters.Add("limit", 0)

        Using localHttpClient As New HttpClient(New HCHandler)
            Dim RequestUri = New pUri("public/resources", parameters)
            Using response As Net.Http.HttpResponseMessage = Await localHttpClient.GetAsync(RequestUri).ConfigureAwait(False)
                Dim result As String = Await response.Content.ReadAsStringAsync()

                If response.StatusCode = Net.HttpStatusCode.OK Then
                    Return result.Jobj.SelectToken("public_key").ToString
                Else
                    ShowError(result)
                End If
            End Using
        End Using
    End Function
#End Region

#Region "PublicUrlToDirectUrl"
    Public Async Function _PublicUrlToDownloadUrl(PublicUrl As Uri) As Task(Of String) Implements ISharing.PublicUrlToDirectUrl
        PublicUrl.Validation(ItemTypeEnum.both)
        Dim parameters = New Dictionary(Of String, String)
        parameters.Add("public_key", System.Net.WebUtility.UrlEncode(PublicUrl.ToString))
        parameters.Add("limit", 0)

        Using localHttpClient As New HttpClient(New HCHandler)
            Dim RequestUri = New pUri("public/resources", parameters)
            Using response As Net.Http.HttpResponseMessage = Await localHttpClient.GetAsync(RequestUri).ConfigureAwait(False)
                Dim result As String = Await response.Content.ReadAsStringAsync()

                If response.StatusCode = Net.HttpStatusCode.OK Then
                    Return result.Jobj.SelectToken("file").ToString
                Else
                    ShowError(result)
                End If
            End Using
        End Using
    End Function
#End Region

#Region "Metadata"
    Public Async Function _Metadata(PublicUrl As Uri) As Task(Of JSON_FileMetadata) Implements ISharing.Metadata
        PublicUrl.Validation(ItemTypeEnum.both)
        Dim parameters = New Dictionary(Of String, String)
        parameters.Add("public_key", System.Net.WebUtility.UrlEncode(PublicUrl.ToString))
        parameters.Add("limit", 0)

        Using localHttpClient As New HttpClient(New HCHandler)
            Dim RequestUri = New pUri("public/resources", parameters)
            Using response As Net.Http.HttpResponseMessage = Await localHttpClient.GetAsync(RequestUri).ConfigureAwait(False)
                Dim result As String = Await response.Content.ReadAsStringAsync()

                If response.StatusCode = Net.HttpStatusCode.OK Then
                    Return JsonConvert.DeserializeObject(Of JSON_FileMetadata)(result, JSONhandler)
                Else
                    ShowError(result)
                End If
            End Using
        End Using
    End Function
#End Region

    '#Region "DownloadFileFromPublicFolder"
    '    Public Async Function GET_DownloadFileFromPublicFolder(PublicFolderUrl As Uri, FilePathToDownload As String, FileSaveDir As String, Optional ReportCls As IProgress(Of ReportStatus) = Nothing, Optional token As Threading.CancellationToken = Nothing) As Task(Of Boolean) Implements IClient.DownloadFileFromPublicFolder
    '        'DownloadUrl = If(String.IsNullOrEmpty(DownloadUrl), (Await POST_GetDownloadUrl(PublicFolderUrl, FilePathToDownload)).href, DownloadUrl)
    '        'Dim Fname = Net.Http.UriExtensions.ParseQueryString(New Uri(DownloadUrl)).[Get]("filename")
    '        Dim meta = Await _Metadata(PublicFolderUrl)

    '        If ReportCls Is Nothing Then ReportCls = New Progress(Of ReportStatus)
    '        ReportCls.Report(New ReportStatus With {.Finished = False, .TextStatus = "Initializing..."})
    '        Try
    '            Dim progressHandler As New Net.Http.Handlers.ProgressMessageHandler(HCHandler)
    '            AddHandler progressHandler.HttpReceiveProgress, (Function(sender, e)
    '                                                                 ReportCls.Report(New ReportStatus With {.ProgressPercentage = e.ProgressPercentage, .BytesTransferred = e.BytesTransferred, .TotalBytes = If(e.TotalBytes Is Nothing, 0, e.TotalBytes), .TextStatus = "Downloading..."})
    '                                                             End Function)
    '            '''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''
    '            Dim localHttpClient As New Net.Http.HttpClient(progressHandler)
    '            localHttpClient.DefaultRequestHeaders.UserAgent.ParseAdd("ISisYandexSDK") '' UserAgent
    '            localHttpClient.DefaultRequestHeaders.ConnectionClose = m_CloseConnection
    '            localHttpClient.Timeout = m_TimeOut
    '            Dim RequestUri = New Uri(DownloadUrl)
    '            '''''''''''''''''will write the whole content to H.D WHEN download completed'''''''''''''''''''''''''''''
    '            Using ResPonse As Net.Http.HttpResponseMessage = Await localHttpClient.GetAsync(RequestUri, Net.Http.HttpCompletionOption.ResponseContentRead, token).ConfigureAwait(False)

    '                If ResPonse.IsSuccessStatusCode Then
    '                    ReportCls.Report(New ReportStatus With {.Finished = True, .TextStatus = (String.Format("[{0}] Downloaded successfully.", Fname))})
    '                Else
    '                    ReportCls.Report(New ReportStatus With {.Finished = True, .TextStatus = ((String.Format("Error code: {0}", ResPonse.ReasonPhrase)))})
    '                End If

    '                ResPonse.EnsureSuccessStatusCode()
    '                Dim stream_ = Await ResPonse.Content.ReadAsStreamAsync()
    '                Dim FPathname As String = String.Concat(OutputDirPath.TrimEnd("\"), "\", Fname)
    '                Using fileStream = New IO.FileStream(FPathname, IO.FileMode.Append, IO.FileAccess.Write)
    '                    stream_.CopyTo(fileStream)
    '                End Using
    '            End Using
    '        Catch ex As Exception
    '            ReportCls.Report(New ReportStatus With {.Finished = True})
    '            If ex.Message.ToString.ToLower.Contains("a task was canceled") Then
    '                ReportCls.Report(New ReportStatus With {.TextStatus = ex.Message})
    '            Else
    '                Throw CType(ExceptionCls.CreateException(ex.Message, ex.Message), YandexException)
    '            End If
    '        End Try
    '    End Function
    '#End Region

#Region "DownloadPublicFile"
    Public Async Function _DownloadPublicFile(PublicUrl As Uri, FileSaveDir As String, Optional ReportCls As IProgress(Of ReportStatus) = Nothing, Optional token As Threading.CancellationToken = Nothing) As Task(Of Boolean) Implements ISharing.DownloadPublicFile
        PublicUrl.Validation(ItemTypeEnum.file)
        Dim meta = Await _Metadata(PublicUrl)

        If ReportCls Is Nothing Then ReportCls = New Progress(Of ReportStatus)
        ReportCls.Report(New ReportStatus With {.Finished = False, .TextStatus = "Initializing..."})
        Try
            Dim progressHandler As New Net.Http.Handlers.ProgressMessageHandler(New HCHandler)
            AddHandler progressHandler.HttpReceiveProgress, (Function(sender, e)
                                                                 ReportCls.Report(New ReportStatus With {.ProgressPercentage = e.ProgressPercentage, .BytesTransferred = e.BytesTransferred, .TotalBytes = If(e.TotalBytes Is Nothing, 0, e.TotalBytes), .TextStatus = "Downloading..."})
                                                             End Function)
            Dim localHttpClient As New HttpClient(progressHandler)
            '''''''''''''''''will write the whole content to H.D WHEN download completed'''''''''''''''''''''''''''''
            Using ResPonse As Net.Http.HttpResponseMessage = Await localHttpClient.GetAsync(New Uri(meta.DownloadUrl), Net.Http.HttpCompletionOption.ResponseContentRead, token).ConfigureAwait(False)

                ResPonse.EnsureSuccessStatusCode()
                Dim stream_ = Await ResPonse.Content.ReadAsStreamAsync()
                Dim FPathname As String = IO.Path.Combine(FileSaveDir, meta.name)
                Using fileStream = New IO.FileStream(FPathname, IO.FileMode.Append, IO.FileAccess.Write)
                    stream_.CopyTo(fileStream)
                End Using

                If ResPonse.IsSuccessStatusCode Then
                    ReportCls.Report(New ReportStatus With {.Finished = True, .TextStatus = (String.Format("[{0}] Downloaded successfully.", meta.name))})
                Else
                    ReportCls.Report(New ReportStatus With {.Finished = True, .TextStatus = ((String.Format("Error code: {0}", ResPonse.ReasonPhrase)))})
                End If
            End Using
        Catch ex As Exception
            ReportCls.Report(New ReportStatus With {.Finished = True})
            If ex.Message.ToString.ToLower.Contains("a task was canceled") Then
                ReportCls.Report(New ReportStatus With {.TextStatus = ex.Message})
            Else
                Throw ExceptionCls.CreateException(ex.Message, ex.Message)
            End If
        End Try
    End Function
#End Region

End Class

File: src/Csharp/YandexDiskSDK/YandexDiskSDK/Interfaces/SharingClient.cs
using Newtonsoft.Json;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Net;
using System.Net.Http;
using System.Threading;
using System.Threading.Tasks;
using YandexDiskSDK.JSON;
using static YandexDiskSDK.utilitiez;

namespace YandexDiskSDK
{
    public class SharingClient : ISharing
    {

        #region "ListAllSharedLinks"
        public async Task<JSON_FolderList> ListAllSharedLinks(ItemTypeEnum Target = ItemTypeEnum.both, Fields Fields = Fields.nothing, PreviewSizeEnum PreviewSize = PreviewSizeEnum.S_150, int Limit = 20, int Offset = 0)
        {
            var parameters = new Dictionary<string, string>();
            if (Fields != Fields.items || Fields != Fields.type)
            {
                parameters.Add("fields", Fields.ToString());
            }

            parameters.Add("limit", Limit.ToString());
            parameters.Add("offset", Offset.ToString());
            parameters.Add("preview_size", stringValueOf(PreviewSize));
            if (Target != ItemTypeEnum.both)
            {
                parameters.Add("type", Target.ToString());
            }

            using (Base.HttpClient localHttpClient = new Base.HttpClient(new Base.HCHandler()))
            {
                var RequestUri = new Base.pUri("resources/public", parameters);
                using (HttpResponseMessage response = await localHttpClient.GetAsync(RequestUri).ConfigureAwait(false))
                {
                    var result = await response.Content.ReadAsStringAsync();
                    if (response.StatusCode == HttpStatusCode.OK)
                    {
                        var fin = new JSON_FolderList();
                        fin.limit = Convert.ToInt32(result.Jobj().SelectToken("limit").ToString());
                        fin.offset = Convert.ToInt32(result.Jobj().SelectToken("offset").ToString());
                        fin._Files = (from c in result.Jobj().SelectToken("items").ToList().Select((i, JSON_FileMetadata) => i).Where(i => i.SelectToken("type").ToString().Equals("file")).Select(i => JsonConvert.DeserializeObject<JSON_FileMetadata>(i.ToString(), Base.JSONhandler)) select c).ToList();
                        fin._Folders = (from c in result.Jobj().SelectToken("items").ToList().Select((i, JSON_FolderMetadata) => i).Where(i => i.SelectToken("type").ToString().Equals("dir")).Select(i => JsonConvert.DeserializeObject<JSON_FolderMetadata>(i.ToString(), Base.JSONhandler)) select c).ToList();
                        return fin;
                    }
                    else
                    {
                        Base.ShowError(result);
                        return null;
                    }
                }
            }
        }
        #endregion

        #region "PublicFolderContents"
        public async Task<JSON_PublicFolder> PublicFolderContents(Uri FolderPublicUrl, string FilePathInsidePublicFolder = null, SortEnum? Sort = null, bool PreviewCrop = true, PreviewSizeEnum PreviewSize = PreviewSizeEnum.S_150, int Limit = 20, int Offset = 0)
        {
            FolderPublicUrl.Validation(ItemTypeEnum.dir);
            return await PublicFolderContents2(FolderPublicUrl.ToString(), FilePathInsidePublicFolder, Sort, PreviewCrop, PreviewSize, Limit, Offset);
        }

        public async Task<JSON_PublicFolder> PublicFolderContents2(string FolderPublicKey, string FilePathInsidePublicFolder = null, SortEnum? Sort = null, bool PreviewCrop = true, PreviewSizeEnum PreviewSize = PreviewSizeEnum.S_150, int Limit = 20, int Offset = 0)
        {
            var parameters = new Dictionary<string, string>();
            parameters.Add("public_key", WebUtility.UrlEncode(FolderPublicKey));
            parameters.Add("path", FilePathInsidePublicFolder);
            parameters.Add("limit", Limit.ToString());
            parameters.Add("offset", Offset.ToString());
            parameters.Add("preview_crop", PreviewCrop.ToString());
            parameters.Add("preview_size", stringValueOf(PreviewSize));
            if (Sort.HasValue) { parameters.Add("sort", Sort.ToString()); }

            using (Base.HttpClient localHttpClient = new Base.HttpClient(new Base.HCHandler()))
            {
                var RequestUri = new Base.pUri("public/resources", parameters);
                using (HttpResponseMessage response = await localHttpClient.GetAsync(RequestUri).ConfigureAwait(false))
                {
                    var result = await response.Content.ReadAsStringAsync();
                    if (response.StatusCode == HttpStatusCode.OK)
                    {
                        var fin = JsonConvert.DeserializeObject<JSON_PublicFolder>(result, Base.JSONhandler);
                        fin.ItemsList._Files = (from c in result.Jobj().SelectToken("_embedded")["items"].ToList().Select((i, JSON_FileMetadata) => i).Where(i => i.SelectToken("type").ToString().Equals("file")).Select(i => JsonConvert.DeserializeObject<JSON_FileMetadata>(i.ToString(), Base.JSONhandler)) select c).ToList();
                        fin.ItemsList._Folders = (from c in result.Jobj().SelectToken("_embedded")["items"].ToList().Select((i, JSON_FolderMetadata) => i).Where(i => i.SelectToken("type").ToString().Equals("dir")).Select(i => JsonConvert.DeserializeObject<JSON_FolderMetadata>(i.ToString(), Base.JSONhandler)) select c).ToList();
                        return fin;
                    }
                    else
                    {
                        Base.ShowError(result);
                        return null;
                    }
                }
            }
        }
        #endregion

        #region "GetDownloadUrlOfFileInPublicFolder"
        public async Task<string> GetDownloadUrlOfFileInPublicFolder(Uri FolderPublicUrl, string FilePathInsidePublicFolder)
        {
            FolderPublicUrl.Validation(ItemTypeEnum.dir);
            return await GetDownloadUrlOfFileInPublicFolder2(FolderPublicUrl.ToString(), FilePathInsidePublicFolder);
        }
        public async Task<string> GetDownloadUrlOfFileInPublicFolder2(string FolderPublicKey, string FilePathInsidePublicFolder)
        {
            var parameters = new Dictionary<string, string>();
            parameters.Add("public_key", WebUtility.UrlEncode(FolderPublicKey));
            parameters.Add("path", FilePathInsidePublicFolder);

            using (Base.HttpClient localHttpClient = new Base.HttpClient(new Base.HCHandler()))
            {
                var RequestUri = new Base.pUri("resources/download", parameters);
                using (HttpResponseMessage response = await localHttpClient.GetAsync(RequestUri).ConfigureAwait(false))
                {
                    var result = await response.Content.ReadAsStringAsync();

                    if (response.StatusCode == HttpStatusCode.OK)
                    { return result.Jobj().SelectToken("href").ToString(); }
                    else
                    {
                        Base.ShowError(result);
                        return null;
                    }
                }
            }
        }
        #endregion

        #region "PublicUrlToPublicKey"
        public async Task<string> PublicUrlToPublicKey(Uri PublicUrl)
        {
            PublicUrl.Validation(ItemTypeEnum.both);
            var parameters = new Dictionary<string, string>();
            parameters.Add("public_key", WebUtility.UrlEncode(PublicUrl.ToString()));
            parameters.Add("limit", "0");

            using (Base.HttpClient localHttpClient = new Base.HttpClient(new Base.HCHandler()))
            {
                var RequestUri = new Base.pUri("public/resources", parameters);
                using (HttpResponseMessage response = await localHttpClient.GetAsync(RequestUri).ConfigureAwait(false))
                {
                    var result = await response.Content.ReadAsStringAsync();
                    if (response.StatusCode == HttpStatusCode.OK)
                    {
                        return result.Jobj().SelectToken("public_key").ToString();
                    }
                    else
                    {
                        Base.ShowError(result);
                        return null;
                    }
                }
            }
        }
        #endregion

        #region "PublicUrlToDirectUrl"
        public async Task<string> PublicUrlToDirectUrl(Uri PublicUrl)
        {
            PublicUrl.Validation(ItemTypeEnum.both);
            var parameters = new Dictionary<string, string>();
            parameters.Add("public_key", WebUtility.UrlEncode(PublicUrl.ToString()));
            parameters.Add("limit", "0");

            using (Base.HttpClient localHttpClient = new Base.HttpClient(new Base.HCHandler()))
            {
                var RequestUri = new Base.pUri("public/resources", parameters);
                using (HttpResponseMessage response = await localHttpClient.GetAsync(RequestUri).ConfigureAwait(false))
                {
                    var result = await response.Content.ReadAsStringAsync();
                    if (response.StatusCode == HttpStatusCode.OK)
                    {
                        return result.Jobj().SelectToken("file").ToString();
                    }
                    else
                    {
                        Base.ShowError(result);
                        return null;
                    }
                }
            }
        }
        #endregion

        #region "Metadata"
        public async Task<JSON_FileMetadata> Metadata(Uri PublicUrl)
        {
            PublicUrl.Validation(ItemTypeEnum.both);
            var parameters = new Dictionary<string, string>();
            parameters.Add("public_key", WebUtility.UrlEncode(PublicUrl.ToString()));
            parameters.Add("limit", "0");

            using (Base.HttpClient localHttpClient = new Base.HttpClient(new Base.HCHandler()))
            {
                var RequestUri = new Base.pUri("public/resources", parameters);
                using (HttpResponseMessage response = await localHttpClient.GetAsync(RequestUri).ConfigureAwait(false))
                {
                    var result = await response.Content.ReadAsStringAsync();
                    if (response.StatusCode == HttpStatusCode.OK)
                    {
                        return JsonConvert.DeserializeObject<JSON_FileMetadata>(result, Base.JSONhandler);
                    }
                    else
                    {
                        Base.ShowError(result);
                        return null;
                    }
                }
            }
        }
        #endregion

        #region "DownloadPublicFile"
        public async Task<bool> DownloadPublicFile(Uri PublicUrl, string FileSaveDir, IProgress<ReportStatus> ReportCls = null, CancellationToken token = default)
        {
            PublicUrl.Validation(ItemTypeEnum.file);
            var meta = await Metadata(PublicUrl);

            if (ReportCls == null) { ReportCls = new Progress<ReportStatus>(); }
            ReportCls.Report(new ReportStatus { Finished = false, TextStatus = "Initializing..." });

            try
            {
                var progressHandler = new System.Net.Http.Handlers.ProgressMessageHandler(new Base.HCHandler());
                progressHandler.HttpReceiveProgress += (sender, e) => { ReportCls.Report(new ReportStatus { ProgressPercentage = e.ProgressPercentage, BytesTransferred = e.BytesTransferred, TotalBytes = e.TotalBytes, TextStatus = "Downloading..." }); };
                using (Base.HttpClient localHttpClient = new Base.HttpClient(progressHandler))
                {
                    using (HttpResponseMessage ResPonse = await localHttpClient.GetAsync(new Uri(meta.DownloadUrl), HttpCompletionOption.ResponseContentRead, token).ConfigureAwait(false))
                    {
                        ResPonse.EnsureSuccessStatusCode();
                        var stream_ = await ResPonse.Content.ReadAsStreamAsync();
                        var FPathname = Path.Combine(FileSaveDir, meta.name);
                        using (Stream FileStrm = new FileStream(FPathname, FileMode.Append, FileAccess.Write))
                        {
                            stream_.CopyTo(FileStrm);
                        }
                        if (ResPonse.IsSuccessStatusCode)
                        {
                            ReportCls.Report(new ReportStatus { Finished = true, TextStatus = $"[{meta.name}] Downloaded successfully." });
                            return true;
                        }
                        else
                        {
                            ReportCls.Report(new ReportStatus { Finished = true, TextStatus = $"Error code: {ResPonse.ReasonPhrase}" });
                            return false;
                        }
                    }
                }
            }
            catch (Exception ex)
            {
                ReportCls.Report(new ReportStatus { Finished = true });
                if (ex.Message.ToString().ToLower().Contains("a task was canceled"))
                {
                    ReportCls.Report(new ReportStatus { TextStatus = ex.Message });
                }
                else
                {
                    throw ExceptionCls.CreateException(ex.Message, 1001);
                }
                return false;
            }
        }
#endregion


    }
}
